miio/push_server/serverprotocol.py

```diff
@@ -52,57 +52,101 @@ def send_ping_ACK(self, host, port):
+        msg_id = msg_value["id"]
+        _LOGGER.debug("<< %s:%s: %s", host, port, msg_value)
 
-            if host not in self.server._registered_devices:
-                _LOGGER.warning(
-                    "Datagram received from unknown device (%s:%s)",
-                    host,
-                    port,
-                )
-                return
+        # Send OK
+        # This result means OK, but some methods return ['ok'] instead of 0
+        # might be necessary to use different results for different methods
+        payload = {"result": 0}
+        self.send_response(host, port, msg_id, token, payload=payload)
+
+        # Parse message
+        action, device_call_id = msg_value["method"].rsplit(":", 1)
+        source_device_id = device_call_id.replace("_", ".")
+
+        callback(source_device_id, action, msg_value.get("params"))
+
+    def _handle_datagram_from_client(self, host: str, port: int, data):
+        """Handle datagram from a regular client."""
+        token = bytes.fromhex(32 * "0")  # TODO: make token configurable?
+        msg = Message.parse(data, token=token)
+        msg_value = msg.data.value
+        msg_id = msg_value["id"]
+
+        _LOGGER.debug(
+            "Received datagram #%s from regular client: %s: %s",
+            msg_id,
+            host,
+            msg_value,
+        )
 
-            token = self.server._registered_devices[host]["token"]
-            callback = self.server._registered_devices[host]["callback"]
+        methods = self.server.methods
+        if msg_value["method"] not in methods:
+            return self.send_error(host, port, msg_id, token, -1, "unsupported method")
 
-            msg = Message.parse(data, token=token)
-            msg_value = msg.data.value
-            msg_id = msg_value["id"]
-            _LOGGER.debug("<< %s:%s: %s", host, port, msg_value)
+        method = methods[msg_value["method"]]
+        if callable(method):
+            try:
+                response = method(msg_value)
+            except Exception as ex:
+                return self.send_error(host, port, msg_id, token, -1, str(ex))
+        else:
+            response = method
 
-            # Send OK
-            self.send_msg_OK(host, port, msg_id, token)
+        return self.send_response(host, port, msg_id, token, payload=response)
 
-            # Parse message
-            action, device_call_id = msg_value["method"].rsplit(":", 1)
-            source_device_id = device_call_id.replace("_", ".")
+    def datagram_received(self, data, addr):
+        """Handle received messages."""
+        try:
+            (host, port) = addr
+            if data == HELO_BYTES:
+                return self.send_ping_ACK(host, port)
 
-            callback(source_device_id, action, msg_value.get("params"))
+            if host in self.server._registered_devices:
+                return self._handle_datagram_from_registered_device(host, port, data)
+            else:
```

Comment:
I am slightly worried that we might exedently send incorrect messages to a real device if that device is for instance not properly disconnected and the server is restarted or other edge cases.
Maybe we schould make a `register_client` function and check against that instead of just handeling everything which is not registerd as a miio_device as a client.

Reply:
The incorrect messages would be encrypted with a hardcoded token that should be different from a real device, so sending a message to the server with some other will simply cause the device not to respond. That is why I don't think it's a non-issue.